target/linux/ramips/dts/mt7621_netis_n6.dts

```diff
@@ -0,0 +1,223 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "netis,n6", "mediatek,mt7621-soc";
+	model = "netis N6";
+
+	aliases {
+		label-mac-device = &gmac0;
+
+		led-boot = &led_power_green;
+		led-failsafe = &led_system_green;
+		led-running = &led_power_green;
+		led-upgrade = &led_system_green;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		key-0 {
+			label = "wps";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+			debounce-interval = <60>;
+		};
+
+		key-1 {
+			label = "reset";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_USB;
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			trigger-sources = <&xhci_ehci_port1>;
+			linux,default-trigger = "usbport";
+		};
+
+		led-1 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WPS;
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		led_system_green: led-2 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_INDICATOR;
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		led-3 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_green: led-4 {
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_POWER;
+			gpios = <&gpio 17 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&gmac0 {
+	nvmem-cells = <&macaddr_factory_7ef20 0>;
+	nvmem-cell-names = "mac-address";
+};
+
+&gmac1 {
+	status = "okay";
+	label = "wan";
+	phy-handle = <&ethphy4>;
+
+	nvmem-cells = <&macaddr_factory_7ef26 0>;
+	nvmem-cell-names = "mac-address";
+};
+
+&ethphy4 {
+	/delete-property/ interrupts;
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x80000>;
+			read-only;
+		};
+
+		partition@80000 {
+			label = "Config";
+			reg = <0x80000 0x80000>;
+		};
+
+		partition@100000 {
+			label = "Factory";
+			reg = <0x100000 0x80000>;
+			read-only;
+
+			nvmem-layout {
+				compatible = "fixed-layout";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				eeprom_factory_0: eeprom@0 {
+					reg = <0x0 0xe00>;
+				};
+
+				macaddr_factory_7ef20: macaddr@7ef20 {
+					reg = <0x7ef20 0x6>;
+				};
+
+				macaddr_factory_7ef26: macaddr@7ef26 {
+					reg = <0x7ef26 0x6>;
+				};
+			};
+		};
+
+		partition@180000 {
+			label = "firmware";
+			reg = <0x180000 0x7e60000>;
```

Comment:
This device have bad block management flash area @0x7800000 (last 8MiB of flash) so you need reduce size of `firmware`  and `ubi` partitions(new size = old size - 0x800000).

Reply:
Just some facts:
1. Stock bootlog:
```
Initializing NMBM ...
Signature found at block 1023 [0x07fe0000]
First info table with writecount 0 found in block 960
Second info table with writecount 0 found in block 963
NMBM has been successfully attached
```

```
[    2.542568] Creating 4 MTD partitions on "mtk-nand":
[    2.547539] 0x000000000000-0x000000080000 : "Bootloader"
[    2.554201] 0x000000080000-0x000000100000 : "Config"
[    2.560293] 0x000000100000-0x000000180000 : "Factory"
[    2.566544] 0x000000180000-0x000007fe0000 : "firmware"
[    3.410846] 2 fit-fw partitions found on MTD device firmware
[    3.416542] 0x000000180000-0x000000560000 : "kernel"
[    3.422800] 0x000000560000-0x000007fe0000 : "rootfs"
[    3.429829] mtd: device 5 (rootfs) set to be root filesystem
[    3.437658] buf:[MTD_INFO_Flag]
[    3.440785] 3 squashfs-split partitions found on MTD device rootfs
[    3.447173] 0x000000fc0000-0x000001000000 : "rootfs_ext"
[    3.453696] 0x000001000000-0x000001020000 : "info"
[    3.459623] 0x000001020000-0x000007fe0000 : "rootfs_data"
```

As can be seen from the log "`firmware`" (and "`rootfs_data`") partitions extends up to `0x7fe0000`. 

2. An excerpt from the original netis N6 dts:
```
                nand@3000 {
                        compatible = "mediatek,mt7621-nfc";
                        reg = <0x3000 0x800>;
                        ecc-engine = <0x0a>;
                        #address-cells = <0x01>;
                        #size-cells = <0x01>;
                        pinctrl-names = "default";
                        pinctrl-0 = <0x0b>;
                        status = "okay";

                        flash@0 {
                                reg = <0x00>;
                                nand-ecc-mode = "hw";

                                partitions {
                                        compatible = "fixed-partitions";
                                        #address-cells = <0x01>;
                                        #size-cells = <0x01>;

                                        partition@0 {
                                                label = "Bootloader";
                                                reg = <0x00 0x80000>;
                                                read-only;
                                        };

                                        partition@80000 {
                                                label = "Config";
                                                reg = <0x80000 0x80000>;
                                        };

                                        partition@100000 {
                                                label = "Factory";
                                                reg = <0x100000 0x80000>;
                                        };

                                        partition@180000 {
                                                label = "firmware";
                                                reg = <0x180000 0x7e60000>;
                                        };
                                };
                        };
                };
```

As you can see "`firmware`" partition extends up to `0x180000` + `0x7e60000` = `0x7fe0000`

So, currently I can't confirm that this device has BBM flash area on the last 8 MiB of the flash.

@DragonBluep you're on first-name terms with nmbm. What do you think about this topic?

Reply:
Just some maths:
960 x 0x20000 = 0x07800000
It's `First info table with writecount 0 found in block 960` as stock bootlogs says.

Reply:
> It's `First info table with writecount 0 found in block 960` as stock bootlogs says.

Math is good. This string above was generated by u-boot, not by stock firmware. 

This is from netis N6 stock firmware bootlog:
```
[    1.687900] Scanning device for bad blocks
[    2.536885] 4 ofpart partitions found on MTD device mtk-nand
[    2.542568] Creating 4 MTD partitions on "mtk-nand":
[    2.547539] 0x000000000000-0x000000080000 : "Bootloader"
[    2.554201] 0x000000080000-0x000000100000 : "Config"
[    2.560293] 0x000000100000-0x000000180000 : "Factory"
[    2.566544] 0x000000180000-0x000007fe0000 : "firmware"
[    3.410846] 2 fit-fw partitions found on MTD device firmware
[    3.416542] 0x000000180000-0x000000560000 : "kernel"
[    3.422800] 0x000000560000-0x000007fe0000 : "rootfs"
[    3.429829] mtd: device 5 (rootfs) set to be root filesystem
[    3.437658] buf:[MTD_INFO_Flag]
[    3.440785] 3 squashfs-split partitions found on MTD device rootfs
[    3.447173] 0x000000fc0000-0x000001000000 : "rootfs_ext"
[    3.453696] 0x000001000000-0x000001020000 : "info"
[    3.459623] 0x000001020000-0x000007fe0000 : "rootfs_data"
```
Can you see anything about nmbm?

This is nmbm-enabled stock firmware bootlog (ASUS RT-AX53U):
```
[    1.439467] Scanning device for bad blocks
[    3.135657] nmbm nmbm: Signature found at block 1023 [0x07fe0000]
[    3.145197] nmbm nmbm: First info table with writecount 0 found in block 960
[    3.162186] nmbm nmbm: Second info table with writecount 0 found in block 963
[    3.169437] nmbm nmbm: NMBM has been successfully attached
[    3.174956] 7 ofpart partitions found on MTD device nmbm
[    3.180239] Creating 7 MTD partitions on "nmbm":
[    3.184855] 0x000000000000-0x0000000e0000 : "Bootloader"
[    3.191744] 0x0000000e0000-0x0000001e0000 : "nvram"
[    3.198116] 0x0000001e0000-0x0000002e0000 : "Factory"
[    3.204692] 0x0000002e0000-0x0000003e0000 : "Factory2"
[    3.211314] 0x0000003e0000-0x0000035e0000 : "Kernel"
[    3.550277] 2 fit-fw partitions found on MTD device Kernel
[    3.555773] 0x0000003e0000-0x0000007a0000 : "linux"
[    3.562179] 0x00000078e794-0x0000035e0000 : "rootfs"
[    3.568940] mtd: device 7 (rootfs) set to be root filesystem
[    3.574652] 0x0000035e0000-0x0000067e0000 : "Kernel2"
[    3.581510] 0x0000067e0000-0x000007800000 : "jffs2"
```

Feel difference. I don't see any reason for enabling nmbm since it isn't enabled in the stock netis N6 firmware. This will also require layout that is different from the stock one (space for NMBM area).

Reply:
>I don't see any reason for enabling nmbm since it isn't enabled in the stock netis N6 firmware. 

No reason to make it wrong because of the stock firmware make it wrong. No difference only if no bad blocks.
For example, `kernel` partition have bad blocks and `nmbm` is active in kernel. Openwrt correctly rewrite `kernel` remapping blocks into `nmbm` table. uboot read remapped blocks instead of bad and start linux kernel correctly. Update successfully complete.
In other case (no `nmbm` in kernel) you have softbricked device in your hands.
>Feel difference.

Reply:
@123serge123 can you provide any real example / evidence (with log) that netis company makes it wrong, besides theoretical reasoning? :) We have many examples devices without nmbm that works fine even with the bad blocks in the kernel area.

Have you already tried contacting netis support and pointing out the issue you suspected?

Reply:
I can make simpler. :)
Negative review reason: YOUR FIRMWARE TOTALLY DESTROY NMBM REMAP TABLE AFTER FIRST BOOT.

Reply:
@123serge123 and again ambitious statements without proofs... 
Your comment does not comply with the formal logic principles. How can you destroy something that is already destroyed by the stock firmware? Please, look at the stock layout in the messages above.

Thanks for your opinion. At the moment I prefer to keep the current code.

Reply:
@123serge123 I checked the stock firmware bootlog just after the flashing and didn't find confirmations that `rootfs_data` (and bmt area too) was erased.

@123serge123, @MaxS0niX considering the following:
1. u-boot is nmbm-enabled
2. formally bmt area belongs to `rootfs_data` in stock but actually doesn't touched / erased / used
I decided to enable nmbm.

Reply:
@csharper2005 
Ok